README.md
<h2 align="center">HAL 2023</h2>
<h4 align="center">A simple script to chat with OpenAI's ChatGPT</h4>
<p align="center"><img src="./image-assets/icon.png" width="150" height="150"><p>
<p align="center">
        <img alt="Script" src="https://img.shields.io/badge/Shell_Script-121011?style=for-the-badge&logo=gnu-bash&logoColor=white">
        <img alt="License" src="https://img.shields.io/badge/MIT-LICENSE-1976D2?style=for-the-badge">
        <img alt="License" src="https://img.shields.io/badge/v-1.10.10-D8DEE9?style=for-the-badge">
</p>

### Index

- [Introduction](#introduction)
- [Getting Started](#getting-started)
- [Screenshots & Demo](#screenshots)
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [Uninstall](#uninstall)
- [Project Structure](#project-structure)
- [Documentation](#documentation)

#### Introduction

Inspired by the infamous HAL9000 from the cult classic "2001: A Space Odyssey", this simple CLI enables you to
interact with [OpenAI](https://openai.com)'s [ChatGPT](https://openai.com/blog/chatgpt/) and generate images.

#### Getting Started

You would need an API key from OpenAI to use this CLI.
Open [this link](https://beta.openai.com/account/api-keys) and generate an API key for yourself.
Upon installation, this CLI stores your personal API key in a file called `.chat-api-key` in your `Home` directory.

This CLI is using the `gpt-3.5-turbo` model by default. You can read more about
it [here](https://platform.openai.com/docs/guides/chat).

To interact with `text-davinci-003` model, simply start your query with `-davinci` followed by a space and the rest of
your query after that.

`gpt-4` Model is also available. Start a prompt with `-gpt4` followed by the rest of the query.

#### Screenshots

<p align="center"><img src="./image-assets/screenshot-1.png"><p>
<p align="center"><img src="./image-assets/screenshot-2.png"><p>
<p align="center"><img src="./image-assets/screenshot-3.png"><p>

#### Requirements

- [curl](https://curl.se/)
- [jq](https://stedolan.github.io/jq/)
- [wget](https://www.gnu.org/software/wget/)

##### Optional 
- [glow](https://github.com/charmbracelet/glow) helps with better output formatting.

#### Installation

- Clone the repo
  ```sh
  git clone https://github.com/Brutuski/hal2023-cli.git
  ```
- Change directory
   ```sh
   cd hal2023-cli
   ```
- Run the script
  ```sh
  make install
  ```

#### Usage

- Run the command `hal2023`


- To exit simply type `exit` or `quit` or `:q`


- Default model is `gpt-3.5-turbo`. If you would like to interact with the `text-davinci-003` model simply start your
  query with `-davinci`

- To interact with `gpt-4` model, start the prompt with `-gpt4` followed by the query.

- Image generation has been added. Simple start your input with `-image ` followed by the image prompt. Generated image
  URL is shortened using the [tny.im API](https://tny.im/aboutapi.php) and displayed.

- Type `-email` or `--e` and press enter for quick email generation process.

- Type `-article` or `--a` and press enter for quick article generation process.

- Type `-cmd` followed by the query to generate single line shell commands ready to be executed in he terminal.


- More options:
  ```sh
  make check
  make about
  make help
  ```
- A history feature has been added as of `v1.4.1`. The history file is created and maintained
  at `~/.hal2023_history.txt`.

#### Uninstall

```sh
make uninstall
```

Please note: The history file (located at ~/.hal2023_history.txt) is not removed while uninstalling. You can delete it
manually if required.

#### Project Structure

```
├── hal2023                     # Main script
├── README.md                   # Documentation file
├── CONTRIBUTING.md             # Contributing document
├── CODE_OF_CONDUCT.md          # Code of Conduct document
├── Makefile                    # To install, uninstall, check requirements and a short about section
├── check_requirements.sh       # Script that checks for requirements
├── image-assets                # Directory for all documentation related image assets
├── .gitignore                  # File to primarily exclude .idea and .DS_Store
├── .github                  
  └── CHANGELOG.md              # Changelog
  ├── ISSUES_TEMPLATE.md        # Issues reporting template
  └── PULL_REQUEST_TEMPLATE.md  # Pull request submission template
└── LICENSE                     # MIT License file
```

#### Documentation

[Issues Template](.github/ISSUES_TEMPLATE.md). Please report any issues or feature requests [here](https://github.com/Brutuski/hal2023-cli/issues).

Changes can be tracked in the [CHANGELOG](.github/CHANGELOG.md)

To contribute, please check out [Contributing](CONTRIBUTING.md) and [Pull Request Template](.github/PULL_REQUEST_TEMPLATE.md)

[Code of Conduct](CODE_OF_CONDUCT.md)

#### License [MIT](https://github.com/Brutuski/hal2023-cli/blob/main/LICENSE)



CHANGELOG.md
<h2 align="center"> Changelog</h2>
<p align="center"><img src="./image-assets/icon.png" width="150" height="150"><p>
<p align="center">
        <img alt="Script" src="https://img.shields.io/badge/Shell_Script-121011?style=for-the-badge&logo=gnu-bash&logoColor=white">
        <img alt="License" src="https://img.shields.io/badge/MIT-LICENSE-1976D2?style=for-the-badge">
</p>

#### v1.0.0

_23/01/2023_

- Initial Release.

#### v1.1.0

_24/01/2023_

- Added a `Makefile` for easy installation.
- Cleaned up the script.

#### v1.2.0

_02/03/2023_

- Added Error Message Handling

#### v1.2.1

_02/03/2023_

- Added 'Check Requirements' to Makefile

#### v1.3.1

_07/03/2023_

- Changed default model to `gpt-3.5-turbo`
- Moved unfinished _history_ feature to `alpha branch`.

#### v1.4.1

_10/03/2023_

- Added a history feature.
    - History file is located at `~/.hal2023_history.txt`

#### v1.5.1

_11/03/2023_

- Added option for interacting with `text-davinci-003`
    - Pass arguments with `-davinci ` followed by the query.

#### v1.6.1

_20/03/2023_

- Added image generation.
    - Pass arguments with `-image ` followed by the image generation prompt.

#### v1.6.2

_22/03/2023_

- Better formatting, following this [source](https://google.github.io/styleguide/shellguide.html#s7-naming-conventions).

#### v1.6.3

_23/03/2023_

- Image URL shortener added. URL shortened using [tny.im API](https://tny.im/aboutapi.php)

#### v1.6.4

_24/03/2023_

- More robust requirements checking.
    - `check_requirements.sh` has been updated to specifically point out which of the required tools is
      missing.

#### v1.7.4

_27/03/2023_

- Added 'Used Tokens' feature. Only available for text based queries
    - Tokens used are also reflected in the history file.
- Minor reformatting for history file to make it visually easier to read.

#### v1.7.5

_27/03/2023_

- Fixed `connection_error` function. It is no longer causing an error message by default.
- In case an error occurs, the specific error message from OpenAI is now displayed.

#### v1.7.6

_27/03/2023_

- Added comprehensive comments to the script

#### v1.8.6

_28/03/2023_

- `gpt-4` Model added

#### v1.8.7

_30/03/2023_

- Readline input mode added. Allows the user to navigate the query they are typing using the arrow keys and correct it
  as required
- Minor formatting

#### v1.9.7

_05/04/2023_

- Added pre-defined prompts to make content generation easier. Prompts for the following have been added:

  - Email `-email` or `--e`
  - Articles `-article` or `--a`
- Further comments added
- Minor formatting
 
#### v1.9.8

_06/04/2023_

- Text folding added for ChatGPT responses. Text folds and line breaks appear according to the width of the terminal.

#### v1.10.8

_07/04/2023_

- Added pre-defined prompt to generate single line shell commands.
  
  - Start your query with `-cmd` 

#### Versioning System

HAL 2023 CLI uses [Semantic versioning 2.0.0](https://semver.org)

#### License [MIT](https://github.com/Brutuski/hal2023-cli/blob/main/LICENSE)


check_requirements.sh
#!/bin/bash

# Checks to see if the required tools mentioned in the array are installed on the system
# Currently only curl, jq and wget are required for HAL 2023
# Passes on the names of the missing tools to the main() function
function check_requirements() {
  local requirements=("curl" "jq" "wget")
  local missing=()

  for req in "${requirements[@]}"; do
    if ! command -v "$req" >/dev/null 2>&1; then
      missing+=("$req")
    fi
  done

  main "${missing[@]}"
}

# Prints either a success message
# Or specifies which tool is missing
# 'glow' is an optional requirement
function main() {
  if [[ $# -eq 0 ]]; then
    echo -e "All requirements are fulfilled."
    echo -e "Run '\033[1msudo make install\033[0m' to install."
    if ! command -v "glow" >/dev/null 2>&1; then
      echo -e "\nOptionally, consider installing 'glow' for better formatted outputs and Markdown rendering"
      echo -e "Source: https://github.com/charmbracelet/glow"
    fi
    exit 0;
  else
    echo -e "The following core requirements are missing: \033[1m${missing[*]:-}\033[0m"
    echo -e "Please install these requirements to use HAL 2023 CLI."
    if ! command -v "glow" >/dev/null 2>&1; then
      echo -e "\nOptionally, consider installing 'glow' for better formatted outputs and Markdown rendering"
      echo -e "Source: https://github.com/charmbracelet/glow"
    fi
    exit 1;
  fi
}

# Kicks off this script
check_requirements

hal2023.sh
#!/bin/bash

## GLOBAL VARIABLES ##

_NAME="HAL 2023 CLI"
_VERSION="v1.0.0"
_SOURCE="https://www.github.com/brutuski/"
_ABOUT="A simple CLI to interract with OpenAI's ChatGPT"

## COLORS ##
BLACK='\033[0;30m'
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[0;33m'
BLUE='\033[0;34m'
PURPLE='\033[0;35m'
CYAN='\033[0;36m'
WHITE='\033[0;37m'
RESET='\033[0m'

## FORMATTING ##
_ASCII=$(cat << EOF
██╗  ██╗ █████╗ ██╗         ██████╗  ██████╗ ██████╗ ██████╗
██║  ██║██╔══██╗██║         ╚════██╗██╔═████╗╚════██╗╚════██╗
███████║███████║██║          █████╔╝██║██╔██║ █████╔╝ █████╔╝
██╔══██║██╔══██║██║         ██╔═══╝ ████╔╝██║██╔═══╝  ╚═══██╗
██║  ██║██║  ██║███████╗    ███████╗╚██████╔╝███████╗██████╔╝
╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝    ╚══════╝ ╚═════╝ ╚══════╝╚═════╝
EOF
)

## API RELATED ##
api_key=$(<api_key.txt)

## FUNCTIONS ##

function about(){
    echo -e "Name:    ${WHITE}$_NAME${RESET}"
    echo -e "Version: ${WHITE}$_VERSION${RESET}"
    echo -e "Source:  ${WHITE}$_SOURCE${RESET}"
    echo -e "About:   ${WHITE}$_ABOUT${RESET}"
    exit 0
}

function help(){
    echo "hal2023 -h -> Help"
    echo "hal2023 -v -> About"
    exit 0
}

while getopts "ha" opt; do
    case "$opt" in
        h) help;;
        a) about;;
        *) exit 1;;
    esac
done

printf "\033c"
echo -e "\n${YELLOW}${_ASCII}${RESET}\n"

while true; do
  echo -e "\n\033[2mType exit, quit or :q to exit.\033[0m"
  echo -e "${WHITE}Write a message...${RESET}"
  read input

  if [ "$input" == "exit" ]; then
      echo -e "\nSuccessfuly quit HAL2023\n"
      exit 0
  elif [ "$input" == "quit" ]; then
      echo -e "\nSuccessfuly quit HAL2023\n"
      exit 0
  elif [ "$input" == ":q" ]; then
      echo -e "\nSuccessfuly quit HAL2023\n"
      exit 0
  else

	response=$(curl https://api.openai.com/v1/completions -sS \
  		-H 'Content-Type: application/json' \
  		-H "Authorization: Bearer $api_key" \
  		-d '{
  			"model": "text-davinci-003",
  			"prompt": "'"${input}"'",
  			"max_tokens": 4000,
  			"temperature": 1.0
	}' | jq -r '.choices[].text' | sed '1d')

	echo -e "\n${CYAN}\033[1mChatGPT:\033[0m${RESET}\n${GREEN}${response}${RESET}"
  fi
done
